onnxruntime/python/onnxruntime_pybind_state.cc

```diff
@@ -905,105 +1012,287 @@ void addObjectMethods(py::module& m, Environment& env) {
+          throw std::runtime_error("Not a valid numpy type");
+        }
+
+        int type_num = dtype->type_num;
+        Py_DECREF(dtype);
+
+        if (!IsNumericNumpyType(type_num)) {
+          throw std::runtime_error("Creation of OrtValues is currently only supported from non-string numpy arrays");
+        }
+
+        auto ml_value = onnxruntime::make_unique<OrtValue>();
+
+        std::unique_ptr<Tensor> tensor;
+        // The tensor's memory is allocated on the CPU
+        if (GetDeviceName(device) == CPU) {
+          tensor = onnxruntime::make_unique<Tensor>(NumpyTypeToOnnxRuntimeType(type_num), shape, GetAllocator());
+        } else if (GetDeviceName(device) == CUDA) {
+      // The tensor's memory is allocated on CUDA
+#ifdef USE_CUDA
+          if (!IsCudaDeviceIdValid(logging::LoggingManager::DefaultLogger(), device.Id())) {
+            throw std::runtime_error("The provided device id doesn't match any available GPUs on the machine.");
+          }
+
+          tensor = onnxruntime::make_unique<Tensor>(NumpyTypeToOnnxRuntimeType(type_num), shape, GetCudaAllocator(device.Id()));
+#else
+        throw std::runtime_error(
+            "Can't allocate memory on the CUDA device using this package of OnnxRuntime. "
+            "Please use the CUDA package of OnnxRuntime to use this feature.");
+#endif
+        } else {
+          throw std::runtime_error("Unsupported device: Cannot place the OrtValue on this device");
+        }
+
+        auto ml_tensor = DataTypeImpl::GetType<Tensor>();
+        ml_value->Init(tensor.release(),
+                       ml_tensor,
+                       ml_tensor->GetDeleteFunc());
+
+        return ml_value;
+      })
+      .def("data_ptr", [](OrtValue* ml_value) -> int64_t {
+        // TODO: Assumes that the OrtValue is a Tensor, make this generic to handle non-Tensors
+        ORT_ENFORCE(ml_value->IsTensor(), "Only OrtValues that are Tensors are currently supported");
+
+        auto* tensor = ml_value->GetMutable<Tensor>();
+
+        if (tensor->Shape().Size() == 0) {
+          return 0;
+        }
+
+        // Should cover x86 and x64 platforms
+        return reinterpret_cast<int64_t>(tensor->MutableDataRaw());
+      })
+      .def("device_name", [](OrtValue* ml_value) -> std::string {
+        // TODO: Assumes that the OrtValue is a Tensor, make this generic to handle non-Tensors
+        ORT_ENFORCE(ml_value->IsTensor(), "Only OrtValues that are Tensors are currently supported");
+
+        return std::string(GetDeviceName(ml_value->Get<Tensor>().Location().device));
+      })
+      .def("shape", [](OrtValue* ml_value) -> py::list {
+        // TODO: Assumes that the OrtValue is a Tensor, make this generic to handle non-Tensors
+        ORT_ENFORCE(ml_value->IsTensor(), "Only OrtValues that are Tensors are currently supported");
+
+        py::list shape_arr;
+        const auto& dims = ml_value->Get<Tensor>().Shape().GetDims();
+
+        for (auto dim : dims) {
```

Comment:
can use std::copy?

Reply:
Well we are adding to a Python list (py::list). At the core, it is still a python object. Not sure if we can employ  std::copy for this .